diff --git a/src/core/or/circuitpadding.h b/src/core/or/circuitpadding.h
index 683ac95de..ea34dacdf 100644
--- a/src/core/or/circuitpadding.h
+++ b/src/core/or/circuitpadding.h
@@ -750,6 +750,11 @@ circpad_decision_t circpad_send_padding_cell_for_callback(
 
 void circpad_free_all(void);
 
+// called by circpad_cell_event_* and circpad_machine_event_* callbacks for the
+// specific event, used to generate traces for circpad simulations.
+MOCK_DECL(void,
+circpad_event_callback, (const char *event, uint32_t circuit_identifier));
+
 #ifdef CIRCUITPADDING_PRIVATE
 STATIC void  machine_spec_free_(circpad_machine_spec_t *m);
 #define machine_spec_free(chan) \
diff --git a/src/test/include.am b/src/test/include.am
index d8e25dea9..650f53f3e 100644
--- a/src/test/include.am
+++ b/src/test/include.am
@@ -118,6 +118,7 @@ src_test_test_SOURCES += \
 	src/test/test_channel.c \
 	src/test/test_channelpadding.c \
 	src/test/test_circuitpadding.c \
+	src/test/test_circuitpadding_sim.c \
 	src/test/test_channeltls.c \
 	src/test/test_checkdir.c \
 	src/test/test_circuitlist.c \
@@ -338,6 +339,7 @@ src_test_test_timers_LDFLAGS = $(src_test_test_LDFLAGS)
 
 # ADD_C_FILE: INSERT HEADERS HERE.
 noinst_HEADERS+= \
+	src/test/circuitpadding_sim_arg.h \
 	src/test/fakechans.h \
 	src/test/hs_test_helpers.h \
 	src/test/log_test_helpers.h \
diff --git a/src/test/test.c b/src/test/test.c
index 6dbec26fa..f2849015b 100644
--- a/src/test/test.c
+++ b/src/test/test.c
@@ -834,6 +834,7 @@ struct testgroup_t testgroups[] = {
   { "checkdir/", checkdir_tests },
   { "circuitbuild/", circuitbuild_tests },
   { "circuitpadding/", circuitpadding_tests },
+  { "circuitpadding_sim/", circuitpadding_sim_tests },
   { "circuitlist/", circuitlist_tests },
   { "circuitmux/", circuitmux_tests },
   { "circuitstats/", circuitstats_tests },
diff --git a/src/test/test.h b/src/test/test.h
index 76c4c0ec7..ac644e15f 100644
--- a/src/test/test.h
+++ b/src/test/test.h
@@ -188,6 +188,7 @@ extern struct testcase_t cell_queue_tests[];
 extern struct testcase_t channel_tests[];
 extern struct testcase_t channelpadding_tests[];
 extern struct testcase_t circuitpadding_tests[];
+extern struct testcase_t circuitpadding_sim_tests[];
 extern struct testcase_t channeltls_tests[];
 extern struct testcase_t checkdir_tests[];
 extern struct testcase_t circuitbuild_tests[];
diff --git a/src/test/testing_common.c b/src/test/testing_common.c
index ff6028ddb..def43a107 100644
--- a/src/test/testing_common.c
+++ b/src/test/testing_common.c
@@ -28,6 +28,7 @@
 #include "lib/crypt_ops/crypto_init.h"
 #include "lib/version/torversion.h"
 #include "app/main/subsysmgr.h"
+#include "test/circuitpadding_sim_arg.h"
 
 #include <stdio.h>
 #ifdef HAVE_FCNTL_H
@@ -283,6 +284,15 @@ main(int c, const char **v)
       loglevel = LOG_DEBUG;
     } else if (!strcmp(v[i], "--accel")) {
       accel_crypto = 1;
+    } else if (!strcmp(v[i], "--circpadsim")) {
+      if (i + 3 > c) {
+        printf("not enough arguments for --circpadsim, expect %s %s, exiting.\n", 
+          v[0], " --circpadsim <client-trace> <relay-trace>");
+        return 1;
+      }
+      circpad_sim_arg_client_trace = v[i+1];
+      circpad_sim_arg_relay_trace = v[i+2];
+      i += 2;
     } else {
       v[i_out++] = v[i];
     }
